Allow composite to be a sub-aggregation of filter

not-napoleon opened the pull request:
This PR is intended to replace #38863.  After this change, composite can be an arbitrarily deep child aggregation of any combination of nested and filter aggregations.

Relates to #37178

not-napoleon committed:
Allow composite to be a sub-aggregation of filters

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nik9000 reviewed (commented):
Could you add something to `CompositeAggregatorTests`? Its nice to have this there too so its easier to use a debugger and stuff.

not-napoleon commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/230_composite.yml:
```diff
@@ -1120,3 +1120,44 @@ setup:
+        rest_total_hits_as_int: true
+        index: verynested
+        body:
+          "aggregations": {
```
I can change it if you feel strongly.  I was copying back and forth between dev console while tinkering with the tests, that's why it's in json.

not-napoleon committed:
ouch, my line count!

nik9000 commented:
The rest compat failure just need master merged in.

I did talk to @not-napoleon about adding a test for when the parent is just a `filter` agg. This is probably a silly thing to do - you can just put the query on top level query instead of the filter agg. But the code he's written allows it. And, surprise! It spits out the wrong results.

not-napoleon commented:
yeah, good call on testing that.  I'm working on a fix for it now.

not-napoleon committed:
Remove nested field from most composite tests

Having the nested field forced the execution path onto the "slow" path,
even for non-nested tests.  This hurt our coverage.  Now we have an
explicitly nested index for testing nested stuff, and a non-nested index
for testing everything else.

not-napoleon committed:
WIP - I don't understand why this passes.

not-napoleon committed:
actually test composite, not terms

not-napoleon committed:
use the slow path when we have parent aggs

not-napoleon commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into composite-nested-filter-2

not-napoleon committed:
BWC skip dance

not-napoleon committed:
Merge branch 'composite-nested-filter-2' of github.com:not-napoleon/elasticsearch into composite-nested-filter-2

nik9000 reviewed (approved):
Makes sense to me.

not-napoleon commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into composite-nested-filter-2

BountyCountry commented:
@not-napoleon I don't see this one in the 7.13.0 release notes. Did it make it?

not-napoleon commented:
It should be there.  I see it in the git log for 7.13, and we would have updated the tags on this issue if we'd bumped it forward.  Not sure why it didn't make the release notes though.
